File: backend/app/models/project.py
import enum
import uuid
from datetime import date, datetime, timezone

from sqlalchemy import Date, DateTime, Enum as SQLEnum, ForeignKey, Numeric, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class ProjectStatus(str, enum.Enum):
    planning = "planning"
    active = "active"
    on_hold = "on_hold"
    completed = "completed"
    canceled = "canceled"


class ProjectPriority(str, enum.Enum):
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"


class Project(Base):
    __tablename__ = "projects"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    company_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("companies.id", ondelete="CASCADE"), nullable=False, index=True
    )
    created_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    client_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    location: Mapped[str | None] = mapped_column(String(500), nullable=True)

    status: Mapped[ProjectStatus] = mapped_column(
        SQLEnum(ProjectStatus, name="projectstatus"), nullable=False, default=ProjectStatus.planning
    )
    priority: Mapped[ProjectPriority] = mapped_column(
        SQLEnum(ProjectPriority, name="projectpriority"), nullable=False, default=ProjectPriority.medium
    )

    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    budget: Mapped[float | None] = mapped_column(Numeric(15, 2), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

File: frontend/src/types/project.ts
export type ProjectStatus = "planning" | "active" | "on_hold" | "completed" | "canceled";
export type ProjectPriority = "low" | "medium" | "high" | "critical";

export interface Project {
  id: string;
  company_id: string;
  created_by_id: string | null;
  name: string;
  description: string | null;
  client_name: string | null;
  location: string | null;
  status: ProjectStatus;
  priority: ProjectPriority;
  start_date: string | null;
  end_date: string | null;
  budget: number | null;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string;
  client_name?: string;
  location?: string;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  start_date?: string;
  end_date?: string;
  budget?: number;
}

export const STATUS_CONFIG: Record<ProjectStatus, { label: string; color: string }> = {
  planning:  { label: "Planning",   color: "bg-blue-100 text-blue-800" },
  active:    { label: "Active",     color: "bg-emerald-100 text-emerald-800" },
  on_hold:   { label: "On Hold",    color: "bg-amber-100 text-amber-800" },
  completed: { label: "Completed",  color: "bg-slate-100 text-slate-700" },
  canceled:  { label: "Canceled",   color: "bg-red-100 text-red-700" },
};

export const PRIORITY_CONFIG: Record<ProjectPriority, { label: string; color: string }> = {
  low:      { label: "Low",      color: "bg-slate-100 text-slate-600" },
  medium:   { label: "Medium",   color: "bg-blue-100 text-blue-700" },
  high:     { label: "High",     color: "bg-orange-100 text-orange-700" },
  critical: { label: "Critical", color: "bg-red-100 text-red-700" },
};
